lib/web/apiserver_test.go

```diff
@@ -8965,6 +8966,74 @@ func TestLogout(t *testing.T) {
 	require.ErrorIs(t, err, trace.AccessDenied("need auth"))
 }
 
+// TestSAMlSessionClearedOnLogout tests if SAML IdP session is cleared on logout.
+func TestSAMlSessionClearedOnLogout(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	env := newWebPack(t, 2)
+
+	const user = "llama"
+	const samlSessionID = "saml_session_id"
+
+	// create logged in session
+	pack := env.proxies[0].authPack(t, user, nil /* roles */)
+
+	// manually add SAML IdP session. The actual SAML IdP session and session cookie is set
+	// by the SAML IdP and the code to do that is on teleport.e.
+	_, err := env.proxies[0].client.CreateSAMLIdPSession(ctx, types.CreateSAMLIdPSessionRequest{
+		SessionID:   samlSessionID,
+		Username:    pack.user,
+		SAMLSession: &types.SAMLSessionData{ID: samlSessionID},
+	})
+	require.NoError(t, err)
+	// add SAML session session cookie to authenticated client pack.
+	jar, err := cookiejar.New(nil)
+	require.NoError(t, err)
+	setSAMLCookie := &http.Cookie{
+		Name:     samlidp.SAMLSessionCookieName,
+		Value:    samlSessionID,
+		MaxAge:   int(time.Second) * 5,
+		HttpOnly: true,
+		Secure:   true,
+		Path:     "/",
+	}
+	jar.SetCookies(&env.proxies[0].webURL, append(pack.cookies, setSAMLCookie))
+	pack2 := env.proxies[0].newClient(t, roundtrip.BearerAuth(pack.session.Token), roundtrip.CookieJar(jar))
+
+	samlSession, err := env.proxies[0].client.GetSAMLIdPSession(ctx, types.GetSAMLIdPSessionRequest{
+		SessionID: samlSessionID,
+	})
+	require.NoError(t, err)
+	require.Equal(t, user, samlSession.GetUser())
+	require.Equal(t, samlSessionID, samlSession.GetSAMLSession().ID)
+
+	// logout from web. The saml session needs to be deleted and the proxy should
+	// respond with SAML session cookie with empty value.
+	resp, err := pack2.Delete(ctx, pack.clt.Endpoint("webapi", "sessions", "web"))
+	require.NoError(t, err)
+	require.True(t, hasEmptySAMLSessionCookieValue(resp.Cookies()))
+	_, err = env.proxies[0].client.GetSAMLIdPSession(ctx, types.GetSAMLIdPSessionRequest{
+		SessionID: samlSessionID,
+	})
+	require.ErrorContains(t, err, `key "/saml_idp/sessions/saml_session_id" is not found`)
+}
+
+func hasEmptySAMLSessionCookieValue(cookies []*http.Cookie) bool {
+	samlCookie := &http.Cookie{
+		Name:     samlidp.SAMLSessionCookieName,
+		Value:    "",
+		Path:     "/",
+		HttpOnly: true,
+		Secure:   true,
+	}
+	for _, cookie := range cookies {
+		if cookie.String() == samlCookie.String() {
```

Comment:
Maybe:

```suggestion
	samlCookie := (&http.Cookie{
		Name:     samlidp.SAMLSessionCookieName,
		Value:    "",
		Path:     "/",
		HttpOnly: true,
		Secure:   true,
	}).String()
	for _, cookie := range cookies {
		if cookie.String() == samlCookie {
```

So we only calculate this string once?